manim/utils/polylabel.py

```diff
@@ -0,0 +1,152 @@
+#!/usr/bin/env python
+from __future__ import annotations
+
+from queue import PriorityQueue
+from typing import TYPE_CHECKING
+
+import numpy as np
+
+if TYPE_CHECKING:
+    from manim.typing import Point3D
+
+
+class Polygon:
+    def __init__(self, rings: list[np.ndarray]) -> None:
+        """
+        Initializes the Polygon with the given rings.
+
+        Parameters
+        ----------
+        rings : list[np.ndarray]
+            List of arrays, each representing a polygonal ring
+        """
+        # Flatten Array
+        csum = np.cumsum([ring.shape[0] for ring in rings])
+        self.array = np.concatenate(rings, axis=0)
+
+        # Compute Boundary
+        self.start = np.delete(self.array, csum - 1, axis=0)
+        self.stop = np.delete(self.array, csum % csum[-1], axis=0)
+        self.diff = np.delete(np.diff(self.array, axis=0), csum[:-1] - 1, axis=0)
+        self.norm = self.diff / np.einsum("ij,ij->i", self.diff, self.diff).reshape(
+            -1, 1
+        )
+
+        # Compute Centroid
+        x, y = self.start[:, 0], self.start[:, 1]
+        xr, yr = self.stop[:, 0], self.stop[:, 1]
+        self.area = 0.5 * (np.dot(x, yr) - np.dot(xr, y))
+        if self.area:
+            factor = x * yr - xr * y
+            cx = np.sum((x + xr) * factor) / (6.0 * self.area)
+            cy = np.sum((y + yr) * factor) / (6.0 * self.area)
+            self.centroid = np.array([cx, cy])
+
+    def compute_distance(self, point: np.ndarray) -> float:
+        """Compute the minimum distance from a point to the polygon."""
+        scalars = np.einsum("ij,ij->i", self.norm, point - self.start)
+        clips = np.clip(scalars, 0, 1).reshape(-1, 1)
+        d = np.min(np.linalg.norm(self.start + self.diff * clips - point, axis=1))
+        return d if self.inside(point) else -d
+
+    def inside(self, point: np.ndarray) -> bool:
+        """Check if a point is inside the polygon."""
+        # Views
+        px, py = point
+        x, y = self.start[:, 0], self.start[:, 1]
+        xr, yr = self.stop[:, 0], self.stop[:, 1]
+
+        # Count Crossings (enforce short-circuit)
+        c = (y > py) != (yr > py)
+        c = px < x[c] + (py - y[c]) * (xr[c] - x[c]) / (yr[c] - y[c])
+        return np.sum(c) % 2 == 1
+
+
+class Cell:
+    def __init__(self, c: np.ndarray, h: float, polygon: Polygon) -> None:
+        """
+        Initializes the Cell, a square in the mesh covering the polygon.
+
+        Parameters
+        ----------
+        c : np.ndarray
+            Center coordinates of the Cell.
+        h : float
+            Half-Size of the Cell.
+        polygon : Polygon
+            Polygon object for which the distance is computed.
+        """
```

Comment:
Instead of adding docstrings to the `__init__` method, the docstring should be directly added to the class.

Same things here:
- Types should not be added again to the docstring.
- `c` should be typed as `npt.NDArray[ManimFloat]`, although at this point we could simply define a `PointND` type alias.

```suggestion
class Cell:
    """A square in a mesh covering the :class:`~.Polygon` passed as an argument.
    
    Parameters
    ----------
        c
            Center coordinates of the cell.
        h
            Half-size of the cell.
        polygon
            :class:`~.Polygon` object for which the distance is computed.
    """
    
    def __init__(self, c: PointND, h: float, polygon: Polygon) -> None:
```